src/scipp/core/data_group.py

```diff
@@ -478,6 +479,36 @@ def _apply_to_items(func: Callable, dgs: Iterable[DataGroup], *args,
          for key in keys})
 
 
+_F = TypeVar('_F', bound=Callable[..., Any])
+
+
+def data_group_overload(func: _F) -> _F:
+    """Add an overload for DataGroup to a function.
+
+    If the first argument of the function is a data group,
+    then the decorated function is mapped over all items.
+    Otherwise, the function is applied directly.
+
+    Parameters
+    ----------
+    func:
+        Function to decorate.
+
+    Returns
+    -------
+    :
+        Decorated function.
+    """
+    # Do not assign '__annotations__' because that causes an error in Sphinx.
+    @wraps(func, assigned=('__module__', '__name__', '__qualname__', '__doc__'))
+    def impl(data, *args, **kwargs):
+        if isinstance(data, DataGroup):
+            return data.apply(lambda x: impl(x, *args, **kwargs))
```

Comment:
So this is operating recursively? This was not clear from the docstring?

Why do we need the lambda? Is `data.apply(impl, *args, **kwargs)` not possible because it is the function being defined currently?